### PR TITLE
CATROID-623 "if( condition , then , else )" Function

### PR DESCRIPTION
**Pull request by Rawbat**
Made the implementation of formulas with more than two parameters possible for this and future formulas by adding an additional list of arbitrary size of children to each formula element.

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaElement.java [hunk 1]
```diff
@@ -215,6 +220,12 @@ private void addFunctionTokens(List<InternToken> tokens, String value, FormulaEl
 			addToken(tokens, FUNCTION_PARAMETER_DELIMITER);
 			tokens.addAll(rightChild.getInternTokenList());
 		}
+		for (int index = 0; index < additionalChildren.size(); index++) {
```
Comment (christophsiegl) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaElement.java:223]: why not using a for loop like in L252? Wouldn't it be less clumsy?

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaElement.java [hunk 2]
```diff
@@ -269,6 +296,11 @@ private boolean containsSpriteInCollision(FormulaElement element, String name) {
 	public final void updateCollisionFormula(String oldName, String newName) {
 		tryUpdateCollisionFormula(leftChild, oldName, newName);
 		tryUpdateCollisionFormula(rightChild, oldName, newName);
+
+		for (FormulaElement child : additionalChildren) {
+			tryUpdateCollisionFormula(child, oldName, newName);
+		}
+
```
Comment (christophsiegl) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaElement.java:303]: L321 in updateCollisionFormulaToVersion: Not sure if this also has to be done for our additional children?

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternFormulaKeyboardAdapter.java [hunk 3]
```diff
@@ -480,6 +485,21 @@
 		return returnList;
 	}
 
+	private List<InternToken> buildTripleParameterFunction(Functions function, InternTokenType firstParameter,
+			String firstParameterNumberValue, InternTokenType secondParameter, String secondParameterNumberValue,
+			InternTokenType thirdParameter, String thirdParameterNumberValue) {
+		List<InternToken> returnList = new LinkedList<InternToken>();
+		returnList.add(new InternToken(InternTokenType.FUNCTION_NAME, function.name()));
```
Comment (christophsiegl) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternFormulaKeyboardAdapter.java:492]: InterTokenType.x is already imported. Please use "FUNCTION_NAME" without "InternTokeType." like in buildDoubleParameterFunction(). Also apply this change to each used InternTokeType in buildTripleParameterFunction().

**Review by christophsiegl — CHANGES_REQUESTED** [review]
Functionality looks good!
Added some little suggestions.
Please extend the language test with a 'NaN' case (as stated in the ticket description) to verify this functionality.
Please add a short summary about your changes to the commit message.

**Review by christophsiegl — APPROVED** [review]
Great!

**Review by wslany — CHANGES_REQUESTED** [review]
Great, but please change the first default parameter to the boolean value "false" (not a string; false can be found under the "Logic" button) instead of 1.

In the Catrobat Language Test testIfThenElseFunction.catrobat, please do not use intermediate variables. Instead, write the formulas directly into the Assert bricks.
image

If you only change the testIfThenElseFunction.catrobat project instead of writing it afresh, please do not forget to delete the variables:
image

**Comment by wslany** [comment]
Also, please increase CURRENT_CATROBAT_LANGUAGE_VERSION in catroid/src/main/java/org/catrobat/catroid/common/Constants.java since the xml structure changes with this function.

**Review by christophsiegl — APPROVED** [review]

**Review by christophsiegl — CHANGES_REQUESTED** [review]
Please resolve merge conflict in Constants.java.

**Review by christophsiegl — APPROVED** [review]

**Review by wslany — APPROVED** [review]

**Comment by wslany** [comment]
Thanks!

**Pull request merged**